lib/psi.rb

```diff
@@ -61,79 +64,256 @@ def append(k)
+      v[:sillconvex]      = @set[p][:sill]            if h[:sill]
+      v[:jamb]            = @set[p][:jamb]            if h[:jamb]
+      v[:jambconcave]     = @set[p][:jamb]            if h[:jamb]
+      v[:jambconvex]      = @set[p][:jamb]            if h[:jamb]
+      v[:headconcave]     = @set[p][:headconcave]     if h[:headconcave]
+      v[:headconvex]      = @set[p][:headconvex]      if h[:headconvex]
+      v[:sillconcave]     = @set[p][:sillconcave]     if h[:sillconcave]
+      v[:sillconvex]      = @set[p][:sillconvex]      if h[:sillconvex]
+      v[:jambconcave]     = @set[p][:jambconcave]     if h[:jambconcave]
+      v[:jambconvex]      = @set[p][:jambconvex]      if h[:jambconvex]
+      v[:corner]          = @set[p][:corner]          if h[:corner]
+      v[:cornerconcave]   = @set[p][:corner]          if h[:corner]
+      v[:cornerconvex]    = @set[p][:corner]          if h[:corner]
+      v[:cornerconcave]   = @set[p][:cornerconcave]   if h[:cornerconcave]
+      v[:cornerconvex]    = @set[p][:cornerconvex]    if h[:cornerconvex]
+      v[:parapet]         = @set[p][:parapet]         if h[:parapet]
+      v[:parapetconcave]  = @set[p][:parapet]         if h[:parapet]
+      v[:parapetconvex]   = @set[p][:parapet]         if h[:parapet]
+      v[:parapetconcave]  = @set[p][:parapetconcave]  if h[:parapetconcave]
+      v[:parapetconvex]   = @set[p][:parapetconvex]   if h[:parapetconvex]
+      v[:party]           = @set[p][:party]           if h[:party]
+      v[:partyconcave]    = @set[p][:party]           if h[:party]
+      v[:partyconvex]     = @set[p][:party]           if h[:party]
+      v[:partyconcave]    = @set[p][:partyconcave]    if h[:partyconcave]
+      v[:partyconvex]     = @set[p][:partyconvex]     if h[:partyconvex]
+      v[:grade]           = @set[p][:grade]           if h[:grade]
+      v[:gradeconcave]    = @set[p][:grade]           if h[:grade]
+      v[:gradeconvex]     = @set[p][:grade]           if h[:grade]
+      v[:gradeconcave]    = @set[p][:gradeconcave]    if h[:gradeconcave]
+      v[:gradeconvex]     = @set[p][:gradeconvex]     if h[:gradeconvex]
+      v[:balcony]         = @set[p][:balcony]         if h[:balcony]
+      v[:balconyconcave]  = @set[p][:balcony]         if h[:balcony]
+      v[:balconyconvex]   = @set[p][:balcony]         if h[:balcony]
+      v[:balconyconcave]  = @set[p][:balconyconcave]  if h[:balconyconcave]
+      v[:balconyconvex]   = @set[p][:balconyconvex]   if h[:balconyconvex]
+      v[:rimjoist]        = @set[p][:rimjoist]        if h[:rimjoist]
+      v[:rimjoistconcave] = @set[p][:rimjoist]        if h[:rimjoist]
+      v[:rimjoistconvex]  = @set[p][:rimjoist]        if h[:rimjoist]
+      v[:rimjoistconcave] = @set[p][:rimjoistconcave] if h[:rimjoistconcave]
+      v[:rimjoistconvex]  = @set[p][:rimjoistconvex]  if h[:rimjoistconvex]
+
+      max = [v[:parapetconcave], v[:parapetconvex]].max
+      v[:parapet] = max unless @has[:parapet]
+      @val[p] = v
```

Comment:
Class member _val_ holds PSI values (in W/K per meter) for all possible variants. As with _has_, this is populated once and held in memory for quick access.